numpy/_core/src/multiarray/array_coercion.c

```diff
@@ -247,13 +247,17 @@ npy_discover_dtype_from_pytype(PyTypeObject *pytype)
 }
 
 /*
- * Note: This function never fails, but will return `NULL` for unknown scalars
- *       and `None` for known array-likes (e.g. tuple, list, ndarray).
+ * Note: This function never fails, but will return `NULL` for unknown scalars or
+ *       known array-likes (e.g. tuple, list, ndarray).
```

Comment:
I guess I could make it always return None, since NULL usually means an error and that might be confusing.

Reply:
Maybe we can make this function public now while we're at it 🤪

Reply:
I suppose we could.  I would be tempted to give an equivalent signature to the `PyDict_GetItemWithError` function (whatever that was again...).

Reply:
Ah that's a good call, I can make it have a signature like `PyDict_GetItemRef` (it's actually used in the PR diff!)

Reply:
(That is, if we make it public, if not yes, this is what I meant...)

But, TBH, I think the Python version of it is more interesting.  Not exactly sure we need it in C.

Reply:
Eh, let's make it public if someone asks for it. I'll merge this once the tests pass.

@hawkinsp please keep on reporting issues as you run into them using the new dtype system.

Reply:
Will do, it's great to see the user dtypes make progress.